cub/cub/device/dispatch/dispatch_merge.cuh

```diff
@@ -0,0 +1,297 @@
+// Odeh et al, "Merge Path - Parallel Merging Made Simple" * doi : 10.1109 / IPDPSW .2012.202
+// The algorithm is the same as AgentPartition for merge sort, but that agent handles a lot more.
+template <typename KeysIt1, typename KeysIt2, typename Offset, typename CompareOp, int items_per_tile>
+CUB_DETAIL_KERNEL_ATTRIBUTES void partition_merge_path_kernel(
+  KeysIt1 keys1,
+  Offset keys1_count,
+  KeysIt2 keys2,
+  Offset keys2_count,
+  Offset num_partitions,
+  Offset* merge_partitions,
+  CompareOp compare_op)
+{
+  const Offset partition_idx = blockDim.x * blockIdx.x + threadIdx.x; // TODO(bgruber): can this be an int?
+  if (partition_idx < num_partitions)
+  {
+    const Offset partition_at       = (cub::min)(partition_idx * items_per_tile, keys1_count + keys2_count);
+    merge_partitions[partition_idx] = cub::MergePath(keys1, keys2, keys1_count, keys2_count, partition_at, compare_op);
+  }
+}
+
+template <typename Agent,
+          typename KeyIt1,
+          typename ValueIt1,
+          typename KeyIt2,
+          typename ValueIt2,
+          typename KeyIt3,
+          typename ValueIt3,
+          typename Offset,
+          typename CompareOp>
+__launch_bounds__(Agent::policy::BLOCK_THREADS) CUB_DETAIL_KERNEL_ATTRIBUTES void merge_kernel(
+  KeyIt1 keys1,
+  ValueIt1 items1,
+  Offset num_keys1,
+  KeyIt2 keys2,
+  ValueIt2 items2,
+  Offset num_keys2,
+  KeyIt3 keys_result,
+  ValueIt3 items_result,
+  CompareOp compare_op,
+  Offset* merge_partitions,
+  vsmem_t global_temp_storage)
+{
+  using vsmem_helper_t = vsmem_helper_impl<Agent>;
+  __shared__ typename vsmem_helper_t::static_temp_storage_t shared_temp_storage;
+  auto& temp_storage = vsmem_helper_t::get_temp_storage(shared_temp_storage, global_temp_storage);
+  Agent{temp_storage.Alias(),
+        THRUST_NS_QUALIFIER::cuda_cub::core::make_load_iterator(typename Agent::policy{}, keys1),
+        THRUST_NS_QUALIFIER::cuda_cub::core::make_load_iterator(typename Agent::policy{}, items1),
+        num_keys1,
+        THRUST_NS_QUALIFIER::cuda_cub::core::make_load_iterator(typename Agent::policy{}, keys2),
+        THRUST_NS_QUALIFIER::cuda_cub::core::make_load_iterator(typename Agent::policy{}, items2),
+        num_keys2,
+        keys_result,
+        items_result,
+        compare_op,
+        merge_partitions}();
+  vsmem_helper_t::discard_temp_storage(temp_storage);
+}
+
+template <typename KeyT, typename ValueT>
+struct device_merge_policy_hub
+{
+  static constexpr bool has_values = !::cuda::std::is_same<ValueT, NullType>::value;
+
+  // TODO(bgruber): is this correct?
+  using tune_type = char[has_values ? sizeof(KeyT) + sizeof(ValueT) : sizeof(KeyT)];
+
+  struct policy300 : ChainedPolicy<350, policy300, policy300>
+  {
+    using merge_policy =
+      agent_policy_t<128,
+                     Nominal4BItemsToItems<tune_type>(7), // TODO(bgruber): merge sort had 11
```

Comment:
This is the correct choice for the type to select policies for? `thrust::merge` used keys and values like this. But CUB's merge sort only uses the key type and ignores the value type.